### PR TITLE
Implement UriResolver.pathToUri()

### PR DESCRIPTION
**Pull request by scheglov**
Prepare for https://dart-review.googlesource.com/c/sdk/+/220480

### CODE REVIEW

File: build_resolvers/lib/src/build_asset_uri_resolver.dart
Comment (jakemac53) [inline build_resolvers/lib/src/build_asset_uri_resolver.dart:194]: nit: add a comment here about why its necessary?

**Review by jakemac53 — COMMENTED** [review]
I assume this is a new method that we will need to implement, but note that this could be breaking for some users even if we have an implementation in build_resolvers pre-released, because they could potentially get a newer analyzer but older build_resolvers for some reason.

Usually the reason for being stuck on an older build_resolvers would be being stuck on an old analyzer for some reason though, so it shouldn't be super common. But some people also likely have build_resolvers pinned to an old version, and would be broken (unless they also pin analyzer).

**Comment by scheglov** [comment]
I will update analyzer to make it backward compatible.
https://dart-review.googlesource.com/c/sdk/+/220480/5/pkg/analyzer/lib/src/generated/source.dart#426

**Comment by jakemac53** [comment]
> I will update analyzer to make it backward compatible.

Ah I thought we just implemented this api but I see we extend it so LGTM

File: build_resolvers/lib/src/build_asset_uri_resolver.dart
Comment (jakemac53) [inline build_resolvers/lib/src/build_asset_uri_resolver.dart:196]: It looks like we always use the `posix` context to create these paths. We should either use the resource providers context there as well or just use posix directly here probably?

Comment (scheglov) [reply build_resolvers/lib/src/build_asset_uri_resolver.dart:196]: There is some contradiction here. The `ResourceProvider` and its `pathContext` should correspond to the style of path(s). If we always build Posix path(s), then `pathContext` must be Posix. Otherwise I don't understand why it would work.

Comment (jakemac53) [reply build_resolvers/lib/src/build_asset_uri_resolver.dart:196]: It seems reasonable to me to update to use the `pathContext` when constructing the paths, do you mind updating that here as well?

Comment (scheglov) [reply build_resolvers/lib/src/build_asset_uri_resolver.dart:196]: Hm... We also use `posix` in `assetPath`, and we don't have access to `ResourceProvider`. In this case, it is better to use `posix` directly here as well.

Comment (jakemac53) [reply build_resolvers/lib/src/build_asset_uri_resolver.dart]: sgtm

**Review by jakemac53 — APPROVED** [review]

**Comment by jakemac53** [comment]
anything else on your end? I pushed up a change to prep for release as well.

**Comment by scheglov** [comment]
Nothing that I can think of.

**Comment by jakemac53** [comment]
ok ill merge this once its happy and then merge internally, and publish when that lands

**Review by natebosch — APPROVED** [review]

**Pull request merged**
